pkg/globalcp/server.go

```diff
@@ -0,0 +1,130 @@
+package globalcp
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/Kong/kuma/pkg/core"
+)
+
+var (
+	globalCPLog = core.Log.WithName("global-cp")
+)
+
+type (
+	GlobalCP interface {
+		Start(<-chan struct{}) error
+	}
+
+	LocalCP struct {
+		URL    string `json:"url"`
+		Active bool   `json:"active"`
+	}
+
+	LocalCPMap map[string]*LocalCP
+
+	GlobalCPPoller struct {
+		sync.RWMutex
+		localCPMap LocalCPMap
+		server     *http.Server
+		newTicker  func() *time.Ticker
+	}
+)
+
+const (
+	tickInterval            = 5 * time.Second
+	globalCPStatisticsAddrt = ":5656"
+)
+
+func NewGlobalCPPoller(localCPList map[string]string) (GlobalCP, error) {
+	mux := http.NewServeMux()
+	poller := &GlobalCPPoller{
+		localCPMap: LocalCPMap{},
+		server: &http.Server{
+			Addr:    globalCPStatisticsAddrt,
+			Handler: mux,
+		},
+		newTicker: func() *time.Ticker {
+			return time.NewTicker(tickInterval)
+		},
+	}
+
+	for name, url := range localCPList {
+		poller.localCPMap[name] = &LocalCP{URL: url, Active: true}
+	}
+
+	mux.HandleFunc("/", poller.StatusHandler)
+	return poller, nil
+}
+
+func (g *GlobalCPPoller) Start(stop <-chan struct{}) error {
+	ticker := g.newTicker()
+	defer ticker.Stop()
+
+	// update the status before running the API
+	g.pollLocalCPs()
+
+	errChan := make(chan error)
+	go func() {
+		err := g.server.ListenAndServe()
+		if err != nil {
+			switch err {
+			case http.ErrServerClosed:
+				globalCPLog.Info("Shutting down server")
+			default:
+				globalCPLog.Error(err, "Could not start an HTTP Server")
+				errChan <- err
+			}
+		}
+	}()
+
+	globalCPLog.Info("starting the Global CP polling")
+	for {
+		select {
+		case <-ticker.C:
+			g.pollLocalCPs()
+		case <-stop:
+			globalCPLog.Info("Stopping down API Server")
+			return g.server.Shutdown(context.Background())
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+func (g *GlobalCPPoller) pollLocalCPs() {
+	for name, localCP := range g.localCPMap {
+		response, err := http.Get(localCP.URL)
+		if err != nil {
+			if localCP.Active {
+				globalCPLog.Info(name + " at " + localCP.URL + " did not respond")
```

Comment:
I find `fmt.Sprintf("%s at %s did not respond", name, localCP.URL)` more readable and GoLand can display it beautifully. But that's up to you